dbms/src/Flash/Coprocessor/DAGContext.cpp

```diff
@@ -173,14 +179,19 @@ void DAGContext::initOutputInfo()
     keep_session_timezone_info = encode_type == tipb::EncodeType::TypeChunk || encode_type == tipb::EncodeType::TypeCHBlock;
 }
 
-String DAGContext::getRootExecutorId()
+void DAGContext::initListBasedExecutors()
 {
-    // If return_executor_id is false, we can get the generated executor_id from list_based_executors_order.
-    return return_executor_id
-        ? root_executor_id
-        : (list_based_executors_order.empty()
-               ? ""
-               : list_based_executors_order.back());
+    if (!return_executor_id)
+    {
+        ExecutorIdGenerator id_generator;
+        traverseExecutorsReverse(dag_request, [&](const tipb::Executor & executor) {
+            const auto & executor_id = id_generator.generate(executor);
+            list_based_executors_order.push_back(executor_id);
+            auto * mutable_executor = const_cast<tipb::Executor *>(&executor);
```

Comment:
a little hack.
how about foreach `dag_request.executors` directly...
for (size_t i = 0; i < dag_request.executors_size(); i++)
{
    const auto & executor_id = id_generator.generate(dag_request.executors[i]);
    blabla
}

Reply:
![img_v2_0e65c29e-3a54-4659-b030-b6ed617d77dg](https://user-images.githubusercontent.com/24507159/226265072-c4ff5f71-e00a-4c1a-8272-66d34d35eef6.jpg)
Although `executors` is const, the elements of `executors` are non-const.

Reply:
> ![img_v2_0e65c29e-3a54-4659-b030-b6ed617d77dg](https://user-images.githubusercontent.com/24507159/226265072-c4ff5f71-e00a-4c1a-8272-66d34d35eef6.jpg) Although `executors` is const, the elements of `executors` are non-const.

dag_request is marked as const. So it's necessary to remove const here.